rust/gui-client/src-tauri/src/client/network_changes/windows.rs

```diff
@@ -112,54 +112,111 @@ pub fn check_internet() -> Result<bool> {
     Ok(have_internet)
 }
 
-/// Worker thread that can be joined explicitly, and joins on Drop
+/// Worker thread (on Windows) that can be joined explicitly, and joins on Drop
 pub(crate) struct Worker {
     inner: Option<WorkerInner>,
     rx: mpsc::Receiver<()>,
 }
 
 /// Needed so that `Drop` can consume the oneshot Sender and the thread's JoinHandle
 struct WorkerInner {
-    thread: std::thread::JoinHandle<Result<(), Error>>,
-    stopper: tokio::sync::oneshot::Sender<()>,
+    thread: thread::JoinHandle<Result<()>>,
+    shutdown_tx: oneshot::Sender<()>,
+}
+
+pub(crate) fn dns_notifier(tokio_handle: tokio::runtime::Handle) -> Result<Worker> {
+    Worker::new("DNS listener", move |tx, shutdown_rx| {
+        // Do `block_on` so we use our own worker thread
+        tokio_handle.block_on(async move {
+            // Use `LocalSet` so we can have the runtime put all our tasks
+            // onto this worker thread, which means they don't need to impl
+            // `Send`.
+            let local = tokio::task::LocalSet::new();
+            local
+                .run_until(async move {
+                    // Use `spawn_local` to keep everything inside the worker thread
+                    tokio::task::spawn_local(dns_worker_task(tx, shutdown_rx)).await
+                })
+                .await
+        })?
+    })
+}
+
+/// Listens for DNS change (on Windows) and signals when we see one
+///
+/// This is meant to be spawned. Cancelling it will cause gaps in DNS listening.
+///
+/// On Windows this must run inside a worker thread because `HANDLE` is `!Send` as of
+/// `windows` 0.58.0.
+async fn dns_worker_task(
+    tx: mpsc::Sender<()>,
+    mut shutdown_rx: oneshot::Receiver<()>,
+) -> Result<()> {
+    let mut dns_listener = async_dns::CombinedListener::new()?;
+    loop {
+        match futures::future::select(&mut shutdown_rx, pin!(dns_listener.notified())).await {
+            Either::Left((_, _)) => break,
+            Either::Right((Err(error), _)) => {
+                tracing::error!(?error, "Error while listening for DNS changes");
+                break;
+            }
+            Either::Right((Ok(()), _)) => {}
+        }
+        tx.send(()).await?;
+    }
+    Ok(())
+}
+
+pub(crate) fn network_notifier(_tokio_handle: tokio::runtime::Handle) -> Result<Worker> {
+    Worker::new("Network listener", |tx, shutdown_rx| {
+        let com = ComGuard::new()?;
+        {
+            let _network_change_listener = Listener::new(&com, tx)?;
+            shutdown_rx.blocking_recv().ok();
```

Comment:
So we need this because Windows doesn't block but instead calls our callback as long as it is alive, right?

And we need to use a thread because the handle returned for that callback is `!Send`.